### PR TITLE
Snowflake Destination - use Azure Blog for source stream stage / copy.

### PR DESCRIPTION
**Pull request by ghost**
## What
#4371
*destination-snowflake*: Allows for more efficient loading to Snowflake on Azure by using the COPY command from and Azure Blob / Storage Integration.

## How
Creating new 'copy' artifacts for *destination-jdbc* following existing patterns for GCS and S3 closely.
Create new / update existing artifacts for *destination-snowflake* to implement new *destination-jdbc* artifacts.
Updated spec.json for *destination-snowflake* to include a new (4th) option for Azure Blob.
Added package reference for azure-blob-storage where needed.
Added new / updated existing tests.
Bumbed versions for both destinations.
Updated documentation for the expanded functionality.

## Recommended reading order
1. `jdbc.copy.azureblob.AzureBlob*.java`
2. `snowflake.SnowflakeAzureBlob*.java`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [x] Connector version bumped like described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</p>
</details>

### CODE REVIEW

**Comment by marcosmarxm** [comment]
Hi @dsdorazio thanks for the contribution! There is a checklist to update a connector, can you check what steps you already made? Also, change from draft to ready to review when you ready. If you need any assistance, let me know!

**Comment by dsdorazio** [comment]
@marcosmarxm All done (for now). There were some bits that I wasn't sure how to perform like the IntegrationTests and bootstrap.md. I suspect that's not relevant because it's a change to an existing Destination? 

I'm happy to go back and perform / update / fix any of those if someone can walk me through what I've wrong.

File: airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json [hunk 1]
```diff
@@ -200,6 +200,65 @@
...
+              "method",
+              "azure_blob_storage_account_name",
+              "azure_blob_storage_container_name",
+              "azure_blob_storage_account_key",
+              "snowflake_azure_external_stage_name"
+            ],
+            "properties": {
+              "method": {
+                "type": "string",
+                "enum": ["Blob Staging"],
```
Comment (sherifnada) [inline airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json:218]: I would recommend changing this to `Azure Blob Staging` to make it clear what blob is referring to

File: docs/integrations/destinations/snowflake.md [hunk 2]
```diff
@@ -189,6 +189,7 @@ Finally, you need to add read/write permissions to your bucket with that email.
 
 | Version | Date      | Pull Request | Subject |
```
Comment (sherifnada) [inline docs/integrations/destinations/snowflake.md:190]: @dsdorazio would you mind adding a section above (similar to the GCS one) for Azure? 

I set one up for testing. here is what I ran on snowflake, for reference: 

```
# Create an integration between snowflake and azure
create storage INTEGRATION azure_airbyte_integration
  TYPE = EXTERNAL_STAGE
  STORAGE_PROVIDER = AZURE
  ENABLED = TRUE
  AZURE_TENANT_ID = '<azure_tenant_id>'
  STORAGE_ALLOWED_LOCATIONS = ('azure://<account_name>.blob.core.windows.net/<blob_container>/');

# Create a stage
create stage azure_airbyte_stage
  url = '<storage_allowed_location_url_above>'
  storage_integration = azure_airbyte_integration;
GRANT USAGE ON integration azure_airbyte_integration TO ROLE AIRBYTE_ROLE;
GRANT USAGE ON stage azure_airbyte_stage TO ROLE AIRBYTE_ROLE;
DESC STORAGE INTEGRATION azure_airbyte_integration;
```

File: airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json [hunk 3]
```diff
@@ -200,6 +200,65 @@
...
+                "description": "A valid key associated with the Azure Storage Account. This is used by Airbyte to write data to the Storage Account, Container, and Blob.",
+                "airbyte_secret": true,
+                "type": "string",
+                "examples": [
+                  "Z8ZkZpteggFx394vm+PJHnGTvdRncaYS+JhLKdj789YNmD+iyGTnG+PV+POiuYNhBg/ACS+LKjd%4FG3FHGN12Nd=="
+                ],
+                "order": 4
+              },
+              "snowflake_azure_external_stage_name": {
+                "title": "Snowflake Azure External Stage",
```
Comment (sherifnada) [inline airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json:255]:
```suggestion
                "title": "Snowflake External Stage Name",
```

**Review by sherifnada — COMMENTED** [review]
@dsdorazio the code generally looks great! not many comments on that front. 

However I am running integration tests locally and many are failing. Could you create a file `airbyte/airbyte-integrations/connectors/destination-snowflake/secrets/copy_azure_blob_config.json` and run the test class `SnowflakeAzureBlobCopyDestinationAcceptanceTest` ? You can run it directly from the IDE for better UX or `./gradlew :airbyte-integrations:connectors:destination-snowflake:integrationTestJava --tests '*SnowflakeAzureBlobCopyDestinationAcceptanceTest'`

File: airbyte-config/init/src/main/resources/seed/destination_definitions.yaml [hunk 4]
```diff
@@ -58,7 +58,7 @@
 - destinationDefinitionId: af7c921e-5892-4ff2-b6c1-4a5ab258fb7e
   name: MeiliSearch
   dockerRepository: airbyte/destination-meilisearch
-  dockerImageTag: 0.2.9
+  dockerImageTag: 0.2.10
```
[the remark this answers is not in the document]
Comment (dsdorazio) [reply airbyte-config/init/src/main/resources/seed/destination_definitions.yaml:61]: @sherifnada I don't think this was from one of my commits, but it definitely wasn't intentional. I wonder if this was a change in master that was merged into my repo, and it was reverted?

**Comment by dsdorazio** [comment]
> @dsdorazio the code generally looks great! not many comments on that front.
> 
> However I am running integration tests locally and many are failing. Could you create a file `airbyte/airbyte-integrations/connectors/destination-snowflake/secrets/copy_azure_blob_config.json` and run the test class `SnowflakeAzureBlobCopyDestinationAcceptanceTest` ? You can run it directly from the IDE for better UX or `./gradlew :airbyte-integrations:connectors:destination-snowflake:integrationTestJava --tests '*SnowflakeAzureBlobCopyDestinationAcceptanceTest'`

@sherifnada I think I'll need some help with one. I'm having difficulty troubleshooting the failures and 'why' the steps are not passing. It looks like the test suite is having difficulty loading the sample files, which would necessarily mean the tests for those dependent on them would fail.

**Comment by marcosmarxm** [comment]
> > @dsdorazio the code generally looks great! not many comments on that front.
> > However I am running integration tests locally and many are failing. Could you create a file `airbyte/airbyte-integrations/connectors/destination-snowflake/secrets/copy_azure_blob_config.json` and run the test class `SnowflakeAzureBlobCopyDestinationAcceptanceTest` ? You can run it directly from the IDE for better UX or `./gradlew :airbyte-integrations:connectors:destination-snowflake:integrationTestJava --tests '*SnowflakeAzureBlobCopyDestinationAcceptanceTest'`
> 
> @sherifnada I think I'll need some help with this one. I'm having difficulty troubleshooting the failures and 'why' the steps are not passing. It looks like the test suite is having difficulty loading the sample files, which would necessarily mean the tests for those dependent on them would fail.

Sorry missed this! Do you need any assistance here?

**Comment by sherifnada** [comment]
@dsdorazio can you post the failure logs you are seeing?

**Comment by dsdorazio** [comment]
> @dsdorazio can you post the failure logs you are seeing?

[snowflake_tests.log](https://github.com/airbytehq/airbyte/files/7280036/snowflake_tests.log)
Log from the following command attached.
`./gradlew :airbyte-integrations:connectors:destination-snowflake:integrationTestJava --tests '*SnowflakeAzureBlobCopyDestinationAcceptanceTest' 2>1 >> snowflake_tests.log`

Thanks for the help!

**Comment by cereusperuv** [comment]
@marcosmarxm What's the current status on this issue? This is a blocker for me and my colleagues right now.

**Comment by marcosmarxm** [comment]
@dsdorazio I'll close this issue due inactivity, let me know if you want to continue working in the future.

**Pull request closed without merging**